README.md
# react-native-e-shop

A simple react native project that consists of:
1) A bottom tabs navigation bar which consists of
   - Home
   - Search
   - Cart
   - Favourites
   - Profile
2) Home tab
   - A `GET` request is triggered with `react-query` to `https://fakestoreapi.com`, which is going to return a list of products.
   - This list is shown in a FlatList in Home tab.
   - The product card contains:
     - Image of product
     - Title
     - Favourite button/indicator
     - `Add to cart` button
     - Price
   - You can interact by pressing a product's card which is going to navigate you to the Products details screen.
   - You can mark as favourite or add the product to your cart from the buttons in the product card.
3) Search tab, is utilizing the same API
   - The search term is not being used in the API call because it is not supported by the selected API
   - The mechanism is ready to use with any API that supports that functionality
   - A mechanism for the last 3 recent searches is available to use
   - After a search an API call will take place through `react-query` which is going to return the list of products
4) Cart tab
   - Here you are going to see a list of products that you have added to cart through the `Add to cart`
   - The card component has a quantity component instead of the `Add to cart` one compared to the home screen cards
   - You can increase the number of each product which is going to change the subtotal of the product and the total cost of your order
   - You can remove an item from your cart by swiping from your right and then tapping the delete button at the right of the product card
   - At the bottom of the screen there is a floating button `Proceed to checkout` which contains the total price of your cart
     - By clicking it you will be navigated to an empty/dummy checkout screen
   - The cart tab button/icon will indicate at the top right the number of products (sum of quantities) in total
5) Favourites tab
   - Contains the list of favourite products
   - You can remove each one by tapping the favourite button in each product card
6) Profile tab
   - Buttons to navigate to screens that use the `@dgreasi/react-native-time-slot-picker` lib to schedule an appointment
7) Product screen, consists of:
   - Image of the product
   - Title and Favourite button section
   - Rating and Votes section
   - Price and `Add to cart` button section
   - Description section which is showing 4 lines of text and has a button to interact with the drawer

## Notes

- `Fakestoreapi` doesn't have pagination, thus no infinite scroll was implemented on home, search lists.
- Favourites in store are in `Object` structure instead of `Array` to make faster the check weather a product is or not favourite
  - Right now the Favourites screen is getting the object and creates a list of products to render.
  - This will not have a correct sort every time for the list of favourites.
  - Nevertheless, the `Object` is the way to go, because in a real life application we are going to get the list of favourites from the server, which is going to be sorted.
  - Thus, right now we are utilizing the Favourites slice `Object` in 2 ways
    - mark a product in whatever list, as favourite
    - get favourite list (unsorted)
- Tested on iOS devices.
- Requests are not working on android when running the app in dev. [Axios issue](https://github.com/axios/axios/issues/5366)

## Environment
```shell
"npm": "8.11.0", # Globally installed
"node": "^16", # Globally installed
"yarn": "1.22.19", # Globally installed
"react": "18.2.0",
"react-native": "0.72.4",
```

## Install dependencies

```shell
nvm use 16
npm install --global yarn # Or use npx before each yarn command

# Install dependencies, start metro bundler and build app
yarn clean-start
yarn ios # Will run application to latest iPhone simulator
```

## App functionality

[<img src="screenshots/1.png" width="180"/>](screenshots/1.png)
[<img src="screenshots/2.png" width="180"/>](screenshots/2.png)
[<img src="screenshots/3.png" width="180"/>](screenshots/3.png)
[<img src="screenshots/4.png" width="180"/>](screenshots/4.png)
[<img src="screenshots/5.png" width="180"/>](screenshots/5.png)
[<img src="screenshots/6.png" width="180"/>](screenshots/6.png)
[<img src="screenshots/7.png" width="180"/>](screenshots/7.png)
[<img src="screenshots/8.png" width="180"/>](screenshots/8.png)
[<img src="screenshots/9.png" width="180"/>](screenshots/9.png)
[<img src="screenshots/10.png" width="180"/>](screenshots/10.png)
[<img src="screenshots/11.png" width="180"/>](screenshots/11.png)
[<img src="screenshots/12.png" width="180"/>](screenshots/12.png)
[<img src="screenshots/13.png" width="180"/>](screenshots/13.png)
[<img src="screenshots/14.png" width="180"/>](screenshots/14.png)
[<img src="screenshots/15.png" width="180"/>](screenshots/15.png)
[<img src="screenshots/16.png" width="180"/>](screenshots/16.png)
[<img src="screenshots/17.png" width="180"/>](screenshots/17.png)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [TODOS]

- (**ANDROID**) Fix `axios` network timeout error ([Issue](https://github.com/axios/axios/issues/5366)).
- (**RN**) Add fast-image for better loading time in lists with images.

## [1.0.3]

### Features

- (**RN**) Add screens to show examples of scheduling an appointment.
- (**RN**) Create dummy hooks to get data for calendar.
- (**RN**) Add mechanism of translations.
- (**RN**) Create extra screen to use only the `@dgreasi/react-native-time-slot-picker` lib.
- (**RN**) Re-structure folder structure.
- (**RN**) Remove unused props from interface.
- (**RN**) Add `react-native-gesture-handler`.
- (**RN**) Add `@dgreasi/react-native-time-slot-picker` and use instead of local component.

## [1.0.2]

### Features

- (**RN**) Update react to `v18.2.0`.
- (**RN**) Update react-native to `v0.71.11`, update native setups for iOS, Android.
- (**RN**) Update react-native-reanimated to `v3.3.0`.
- (**RN**) Update react-native-safe-area-context to `v4.5.3`.
- (**RN**) Update react-native-screens to `v3.21.0`.

## [1.0.1]

### Features

- (**RN**) Update to react `v18.0.0`
- (**RN**) Update to react-native `v0.69.11`, update android, iOS native project
- (**RN**) Update to react-reanimated `v2.9.0`

## [1.0.0]

### Features

- (**RN**) Hide entity for searches. Reset hidden entities with button.
- (**RN**) Monitor network and show appropriate message.
- (**RN**) Add recent searches functionality.
- (**RN**) Add Search screen.
- (**RN**) Add Entity screen.
- (**RN**) Add Home screen.
- (**RN**) Add @react-native-community/netinfo.
- (**RN**) Add axios.
- (**RN**) Add react-native-vector-icons.
- (**RN**) Add react-query.
- (**RN**) Add react-native-flash-message for toast messages.
- (**RN**) Add react-native-config for environment config.
- (**RN**) Fix bug of reanimated and react-navigation on android.
- (**RN**) Add @shopify/restyle for theming.
- (**RN**) Add @react-native-async-storage/async-storage.
- (**RN**) Add react-native-device-info.
- (**RN**) Infinite scroll view with dummy data on redux.
- (**RN**) Add styled-components.
- (**RN**) Add react-native-reanimated for animations.
- (**RN**) Setup basic navigation.
- (**RN**) Setup redux toolkit with 2 basic slices (core, posts).
- (**RN**) Install flipper for react, network, redux debugging.
- (**RN**) Create base screens 'Home', 'Post', 'Profile'.
- (**RN**) Create ImageWithFallback component.
- (**RN**) Create Screen component for screen wrapper.
- (**RN**) Create base api with axios.
- (**RN**) Add lint-staged and husky for pre-commit.
- (**RN**) Add eslint and prettier.
- (**RN**) Init react native typescript project.


src/utils/alert.ts
import { Platform } from 'react-native';
import { MessageOptions, showMessage } from 'react-native-flash-message';
import theme from '~theme/theme';

const getBackgroundColor = (type: any) => {
  if (type === 'info') return theme.colors.secondary400;
};

export const ShowAlert = (options: MessageOptions) => {
  const { message, description, type, position } = options;
  const backgroundColor = getBackgroundColor(type);

  const messageOptions: MessageOptions = {
    message,
    description,
    type,
    titleStyle: {
      fontSize: 15,
      lineHeight: 20,
      ...Platform.select({
        ios: {
          fontWeight: '600',
        },
        android: {
          fontFamily: 'Roboto-Medium',
        },
      }),
    },
    textStyle: {
      fontSize: 13,
      lineHeight: 20,
      ...Platform.select({
        ios: {
          fontWeight: '400',
        },
        android: {
          fontFamily: 'Roboto-Regular',
        },
      }),
    },
    position,
    backgroundColor,
  };

  const style = position === 'bottom' ? { paddingTop: theme.spacing.m } : null;
  if (style) messageOptions.style = style;
  showMessage(messageOptions);
};
